Add ProtT5-XL-BFD model card (#7606)
huggingface/transformers: model_cards/Rostlab/prot_t5_xl_bfd/README.md

---
language: protein
tags:
- protein language model
datasets:
- BFD
---

# ProtT5-XL-BFD model

Pretrained model on protein sequences using a masked language modeling (MLM) objective. It was introduced in
[this paper](https://doi.org/10.1101/2020.07.12.199554) and first released in
[this repository](https://github.com/agemagician/ProtTrans). This model is trained on uppercase amino acids: it only works with capital letter amino acids.

## Model description

> ProtT5-XL-BFD is based on T5 model which pretrained on a large corpus of protein sequences in a self-supervised fashion.

Reviewer: ```suggestion
ProtT5-XL-BFD is based on the T5 model that is pretrained on a large corpus of protein sequences in a self-supervised fashion.
```

This means it was pretrained on the raw protein sequences only, with no humans labelling them in any way (which is why it can use lots of
publicly available data) with an automatic process to generate inputs and labels from those protein sequences.

> One important difference between our T5 model and the original T5 version is the denosing objective.

Reviewer: ```suggestion
One important difference between this T5 model and the original T5 version is the denosing objective.
```

> In the original T5 paper they used span denosing for the T5-3B model, in our case we used Bart like MLM denosing objective.

Reviewer: ```suggestion
The original T5-3M model was pretrained using a span denosing objective, while this model was pre-trained with a Bart-like MLM denosing objective.
```

> The masking probability follows the original T5 training with randomly masks 15% of the amino acids in the input. 

Reviewer: ```suggestion
The masking probability is consistent with the original T5 training by randomly masking 15% of the amino acids in the input.
```

> At the end, the feature extracted from this model revealed that the LM-embeddings from unlabeled data (only protein sequences) captured important biophysical properties governing protein

Reviewer: ```suggestion
It has been shown that the features extracted from this self-supervised model (LM-embeddings) captured important biophysical properties governing protein shape.

shape.
This implied learning some of the grammar of the language of life realized in protein sequences.

## Intended uses & limitations

The model could be used for protein feature extraction or to be fine-tuned on downstream tasks.

> We have noticed in some tasks you could gain more accuracy by fine-tuning the model rather than using it as a feature extractor.

Reviewer: ```suggestion
We have noticed in some tasks on can gain more accuracy by fine-tuning the model rather than using it as a feature extractor.
```

> We have also noticed that for feature extraction, its better to use the feature extracted from the encoder not the decoder.

Reviewer: ```suggestion
We have also noticed that for feature extraction, its better to use the feature extracted from the encoder not from the decoder.
```

### How to use

> Here is how to use this model to get the features of a given protein sequence in PyTorch:

Reviewer: ```suggestion
Here is how to use this model to extract the features of a given protein sequence in PyTorch:
```

sequences_Example = ["A E T C Z A O","S K T Z P"]

sequences_Example = [re.sub(r"[UZOB]", "X", sequence) for sequence in sequences_Example]

ids = tokenizer.batch_encode_plus(sequences_Example, add_special_tokens=True, padding=True)

input_ids = torch.tensor(ids['input_ids'])
attention_mask = torch.tensor(ids['attention_mask'])

with torch.no_grad():
    embedding = model(input_ids=input_ids,attention_mask=attention_mask,decoder_input_ids=None)

# For feature extraction we recommend to use the encoder embedding
encoder_embedding = embedding[2].cpu().numpy()
decoder_embedding = embedding[0].cpu().numpy()
```

## Training data

The ProtT5-XL-BFD model was pretrained on [BFD](https://bfd.mmseqs.com/), a dataset consisting of 2.1 billion protein sequences.

## Training procedure

### Preprocessing

The protein sequences are uppercased and tokenized using a single space and a vocabulary size of 21. The rare amino acids "U,Z,O,B" were mapped to "X".
The inputs of the model are then of the form:

```
Protein Sequence [EOS]
```

The preprocessing step was performed on the fly, by cutting and padding the protein sequences up to 512 tokens.

> The details of the masking procedure for each sequence followed the original T5 model for token masking as following:

Reviewer: ```suggestion
The details of the masking procedure for each sequence are as follows:
```
